svgnative/ports/cairo/CairoSVGRenderer.cpp

```diff
@@ -0,0 +1,437 @@
+    // https://www.cairographics.org/samples/gradient/
+    switch (gradient.type)
+    {
+    case GradientType::kLinearGradient:
+        *pat = cairo_pattern_create_linear( gradient.x1, gradient.y1,
+                                            gradient.x2, gradient.y2 );
+        break;
+
+    case GradientType::kRadialGradient:
+        *pat = cairo_pattern_create_radial( gradient.fx, gradient.fy, 0,
+                                            gradient.cx, gradient.cy, gradient.r );
+        break;
+    }
+
+    // set transform matrix
+    if (gradient.transform)
+        cairo_pattern_set_matrix ( *pat, &(static_cast<CairoSVGTransform*>(gradient.transform.get())->mMatrix) );
+
+    // set "stop"s of gradient 
+    for (const auto& stop : gradient.colorStops)
+    {
+        // here, ColorStop is a pair of offset (in float) and color
+        const auto& stopOffset = stop.first;
+        const auto& stopColor = stop.second;
+
+        cairo_pattern_add_color_stop_rgba( *pat, stopOffset,
+                                           stopColor[0],
+                                           stopColor[1],
+                                           stopColor[2],
+                                           opacity * stopColor[3] );
+    }
+
+    // set the mode how to fill the wide area by a small pattern
+    switch (gradient.method)
+    {
+    case SpreadMethod::kReflect:
+        cairo_pattern_set_extend (*pat, CAIRO_EXTEND_REFLECT );
+        break;
+    case SpreadMethod::kRepeat:
+        cairo_pattern_set_extend (*pat, CAIRO_EXTEND_REPEAT );
+        break;
+    case SpreadMethod::kPad:
+        cairo_pattern_set_extend (*pat, CAIRO_EXTEND_PAD );
+        break;
+    default:
+        cairo_pattern_set_extend (*pat, CAIRO_EXTEND_NONE );
+        break;
+    }
+    return;
+}
+
+inline void appendCairoSvgPath( cairo_t* mCairo, const Path& path )
+{
+    cairo_path_t* cairoPath = getPathObjFromCairoSvgPath(&path);
+    cairo_append_path (mCairo, cairoPath);
+    cairo_path_destroy( cairoPath );
+}
+
+void CairoSVGRenderer::DrawPath(
+    const Path& path, const GraphicStyle& graphicStyle, const FillStyle& fillStyle, const StrokeStyle& strokeStyle)
+{
+    SVG_ASSERT(mCairo);
+    Save(graphicStyle);
+
+    if (fillStyle.hasFill)
+    {
+        if (fillStyle.paint.type() == typeid(Gradient)) {
+            cairo_pattern_t* pat;
+            createCairoPattern(fillStyle.paint, fillStyle.fillOpacity * graphicStyle.opacity, &pat);
+            cairo_set_source(mCairo, pat);
+        } else {
+            const auto& color = boost::get<Color>(fillStyle.paint);
+            cairo_set_source_rgba(mCairo,
+                                  color[0],
+                                  color[1],
+                                  color[2],
+                                  color[3] * fillStyle.fillOpacity * graphicStyle.opacity );
+        }
+
+        // Skia backend does not handle fillStyle.Rule yet
+        switch (fillStyle.fillRule) {
```

Comment:
This could go into another inline function and called for filling and clipping.

Reply:
Indeed, I would introduce an inline function setCairoFillAndClipRule() and use it twice.